### PR TITLE
Migrate GitHub Actions

### PR DESCRIPTION
**Pull request by nd-02110114**
This PR migrates Travis CI to Github Actions, a part of #2289

Please see this PR https://github.com/nd-02110114/deepchem/pull/7 to confirm how this PR works

## Remaining task

- [x] Add PYPI_USERNAME and  PYPI_PASSWORD  to secret env viriables. 
- [x] Link Codecov with this repository

And, I can't test publishing PyPI, so we may face some errors about publishing PyPI steps.

## New CI

I created 4 CI configurations.

- `.github/workflows/main.yml` 
- `.github/workflows/doc.yml` 
- `.github/workflows/build.yml`
- `.github/workflows/release.yml`

###  main.yml

This workflow tests the DeepChem core. This workflow check yapf, flake8, doctest, pytest, upload coverages and publish the nightly build. **The new CI uses the Codecov for checking coverage because the coveralls action is not well maintained** (see: https://github.com/coverallsapp/github-action/issues/30). And then, **the new CI will publish the nightly build package when passing all tests in master branch**. The platform which is tested is Ubuntu (Python 3.6, 3.7), MacOS (Python 3.7) Windows (Python 3.7). Previously, we don't test MacOS, so I added it.

In addition to this, I refactor how to manage dependencies. This refactor is based on the Open Catalyst Project built by Facebook AI. The new approach is here.

1. Create `env.yml` which suits your situation by using conda-merge and base yaml files like `env.common.yml`, `env.cpu.yml`, `env.gpu.yml`.
2. Update the deepchem environment by `conda env update --file env.yml`

### doc.yml

This workflow tests the documents build and doctest.

### build.yml

This workflow tests the build scripts, `install_deepchem_conda.sh` and  `install_deepchem_conda.ps1`. Previously, we don't test `install_deepchem_conda.ps1`, so I added this workflow.

### release.yml

If we push the tag, this workflow will run. This workflow is only publishing the stable PyPI package.

### Vina Installation is not working on Windows

I think `VinaPoseGenerator` is not working on Windows. `msiexec /i .....` is only working on the command prompt. If users use the Powershell or Git bash, the command did not work. And then, if users use the command prompt, we must need the GUI operations, so Vina can't be installed by only CUI operations.

https://github.com/deepchem/deepchem/blob/ab07c0b6b070d376039f55762e6894df289b7364/deepchem/dock/pose_generation.py#L143-L145

### Installation failure in Python 3.8

I tried to add Ubuntu Python 3.8 in this PR, but the Installation was failed. The error indicated the `RDKit`, `openmm` and `pdbfixer` conflicted when installing. Therefore, I don't add python 3.8

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
@nd-02110114 Thank you for kicking this off! This is super valuable to make sure we don't run into serious issues once our Travis support shuts down :)

I did a first pass through the code and am still digesting the structure of it since I'm new to Github actions. At first look, I think everything looks reasonable.

CC @peastman who might have an idea how to handle the `msiexec` failure.

CC @ncfrey @seyonechithrananda as well who might be interested in taking a look

**Comment by peastman** [comment]
> I think VinaPoseGenerator is not working on Windows. msiexec /i ..... is only working on the command prompt.

What happens when it tries to run that command?  Is there an error message?

**Comment by nd-02110114** [comment]
> What happens when it tries to run that command? Is there an error message?

When I tried to run `msiexec /i http://vina.scripps.edu/download/autodock_vina_1_1_2_win32.msi` in Git Bash of my local Windows, it just showed the help window like this. There is no error message.

スクリーンショット 2020-11-15 1 49 33

When running `msiexec /i http://vina.scripps.edu/download/autodock_vina_1_1_2_win32.msi` in Git Bash of  Windows VM which GitHub Actions workflow created, it didn't return anything and the command never ended. (I needed to end the command manually)

About the Powershell, `msiexec /i http://vina.scripps.edu/download/autodock_vina_1_1_2_win32.msi` is not valid.
We need to replace `msiexec` to `msiexec.exe`.

**Comment by peastman** [comment]
This may be an issue with the different treatment of `\` and `/` in bash versus Windows.  Perhaps you need to replace `/i` with `//i`?  Or maybe `'/i'`?

**Comment by nd-02110114** [comment]
> This may be an issue with the different treatment of \ and / in bash versus Windows. Perhaps you need to replace /i with //i? Or maybe '/i'?

When replacing `/i` to `//i`, the command works well. Thanks!

So, the different install command is needed on each platform of Windows. These commands works well.

On Command Prompt
`msiexec /i autodock_vina_1_1_2_win32.msi` 

On Git Bash
`msiexec //i autodock_vina_1_1_2_win32.msi` 

On Powershell
`autodock_vina_1_1_2_win32.msi`

How do we treat these...? I don't come up with any idea... 
And then, when calling these using python's subprocess function, 
the command on Powershell and Git Bash doesn't work. They showed just help windows.

**Comment by peastman** [comment]
For what purpose do we need to handle all three of them?  When running CI, I assume we know what shell it will be running in?  And when Python code is executing it, that will of course use the default Windows shell, not Powershell or Git Bash.

**Comment by nd-02110114** [comment]
> And when Python code is executing it, that will of course use the default Windows shell, not Powershell or Git Bash.

Ah, I misunderstood this point. I didn't understand the Python subprocess is always launching cmd.exe.
We don't need to handle them.

@rbharath 
I think the Vina installation issue was resolved. It is reasonable to skip the Vina installation on Windows CI. This is because the Vina installation need a GUI operation on Windows. And the codes in `pose_generation.py` are not problem. I reverted my modification. This PR is ready to a second review. 

If it is ready to merge, I want to do two things before merging in. Currently, I can't do these things. I need some access authorizations for PyPI and GitHub.

- Add PYPI_USERNAME and PYPI_PASSWORD to secret env viriables.
- Link Codecov with this repository

File: deepchem/molnet/load_function/molnet_loader.py
Comment (rbharath) [inline deepchem/molnet/load_function/molnet_loader.py:50]: Would we have the rdkit import error here now that we no longer have the try/catch block?

**Review by rbharath — COMMENTED** [review]
Looking good! One minor question about try/catch and rdkit on my end

File: deepchem/molnet/load_function/molnet_loader.py
Comment (peastman) [reply deepchem/molnet/load_function/molnet_loader.py:50]: This change is incorrect.  QM7, QM8, and QM9 each require a different size Coulomb matrix.  That's why it can't be specified with a string.

The CoulombMatrix constructor throws an exception if RDKit isn't installed, but not for any good reason.  The constructor doesn't use anything from RDKit.  We could move that check into the `coulomb_matrix()` method, which is the only place it gets used.

File: deepchem/utils/data_utils.py
Comment (peastman) [inline deepchem/utils/data_utils.py:154]: I looked up this URL.  Someone did post this code, but other better solutions have a lot more votes.  According to the top voted answer, all you need to do is add the option `mode='w'` to the NamedTemporaryFile constructor?

File: deepchem/molnet/load_function/molnet_loader.py
Comment (nd-02110114) [reply deepchem/molnet/load_function/molnet_loader.py]: I fixed.

File: deepchem/utils/data_utils.py
Comment (nd-02110114) [reply deepchem/utils/data_utils.py:154]: > all you need to do is add the option mode='w' to the NamedTemporaryFile constructor?

No. The line that happens error is here
https://github.com/deepchem/deepchem/blob/6c5a5405acea333ee7a65a798ddb5c9df702a0b8/deepchem/data/data_loader.py#L432
This line has already received `mode='w'` option.

**Comment by nd-02110114** [comment]
I fixed some codes by following your review. And, I added issue templates (based on PyTorch Geometric), PR template and note which explains how to release. 

If you think this PR is ready to merge, I'll set up PyPI and CodeCov and merge in the master in the next week.

File: notes/relaease.md
Comment (nd-02110114) [inline notes/relaease.md:1]: This note is a still WIP, but I wrote up all things which I know about the release.

File: .github/CONTRIBUTING.md
Comment (rbharath) [inline .github/CONTRIBUTING.md:26]: Should we keep this text around about numpy documentation style here or is it already well covered in the coding conventions page?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (rbharath) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: This is great!

File: docs/source/development_guide/coding.rst
Comment (rbharath) [inline docs/source/development_guide/coding.rst:27]: Ah this is great!

File: notes/relaease.md
Comment (rbharath) [inline notes/relaease.md:1]: Slight typo, `release.md`.

Comment (rbharath) [reply notes/relaease.md:1]: This is great! Thank you so much for putting this together :)

**Review by rbharath — COMMENTED** [review]
This is looking really good! I think it's basically ready to merge in. Will punt over to @peastman first though in case he has additional comments

File: deepchem/molnet/load_function/molnet_loader.py
Comment (peastman) [inline deepchem/molnet/load_function/molnet_loader.py:50]: Why the change from objects to lambdas?

File: docs/source/development_guide/coding.rst
Comment (peastman) [inline docs/source/development_guide/coding.rst:37]: I don't understand what this sentence means?

**Comment by peastman** [comment]
I posted a couple of specific minor questions.  Aside from those, a general issue is that this PR contains a *lot* of changes that seem unrelated to Github Actions.  For this PR it's maybe too late to worry about, but in the future it's much better to split up things like this so each PR only contains related changes.  It makes them easier to review, and also keeps the revision history cleaner.

**Comment by rbharath** [comment]
I think Travis CI is accelerating their shutdown. We should see if we can get this PR merged in sooner rather than later. 

@nd-02110114 Would you have any bandwidth to take a crack at getting this PR ready to merge in?

**Comment by nd-02110114** [comment]
Sorry for stopping the migration work.
I just finished presenting on the domestic conference today. So, I have enough time to work in this weekend. I will reflect your review to this PR and will be ready to merge  in early next week. 
And, I will split  the modification for Issue and PR templates to the another PR.

How do you think...?

**Comment by rbharath** [comment]
@nd-02110114 That sounds like a great idea! Splitting into 2 PRs should make this easier to merge in :). Thanks for taking this project on!

**Comment by nd-02110114** [comment]
I think I finished reflecting your review to the PR and splitting the PR. 

But, in my work, I found we can't install all packages by the new pip resolver (>=20.3). If we install all packages with new pip resolver, the error like https://github.com/pypa/pip/issues/9215 will happen.

The above error happens in OSS like [airflow](https://github.com/apache/airflow) has a lot of dependencies. DeepChem also have a lot of dependencies, so I think we meet such an error.

My workaround is using pip 20.2, but we need to fix this error in the future. 

PS. In Travis CI, we use pip 20.1, so we haven't recognized this issue.

**Review by rbharath — APPROVED** [review]
Thank you for the fixes and the updates @nd-02110114! This is really awesome :). I've just approved Codecov for the DeepChem github org.

I think this PR is ready to merge from my end. I'll mark it as approved but hold off on merging in case @peastman has further comments.

**Comment by nd-02110114** [comment]
I'll merge this PR on the weekend when I can work on it. 
This is because it can cause problems that are not known until after merging, 
and I want to be able to respond to them quickly.

**Comment by nd-02110114** [comment]
I merge in! Goodbye, Travis CI✋

**Pull request merged**
